File: scaffold/src/main/java/top/xuqingquan/web/agent/WebParentLayout.java
package top.xuqingquan.web.agent;

import android.app.Activity;
import android.content.Context;
import android.graphics.Color;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewStub;
import android.widget.FrameLayout;
import androidx.annotation.IdRes;
import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import top.xuqingquan.R;
import top.xuqingquan.utils.Timber;

public class WebParentLayout extends FrameLayout implements Provider<AbsAgentWebUIController> {
    private AbsAgentWebUIController mAgentWebUIController = null;
    @LayoutRes
    private int mErrorLayoutRes;
    @IdRes
    private int mClickId = -1;
    private View mErrorView;
    private android.webkit.WebView mWebView;
    private com.tencent.smtt.sdk.WebView mX5WebView;
    private FrameLayout mErrorLayout = null;

    public WebParentLayout(@NonNull Context context) {
        this(context, null);
        Timber.i("WebParentLayout");
    }

    WebParentLayout(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, -1);
    }

    WebParentLayout(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        if (!(context instanceof Activity)) {
            throw new IllegalArgumentException("WebParentLayout context must be activity or activity sub class .");
        }
        this.mErrorLayoutRes = R.layout.agentweb_error_page;
    }

    public void bindController(AbsAgentWebUIController agentWebUIController) {
        this.mAgentWebUIController = agentWebUIController;
        this.mAgentWebUIController.bindWebParent(this, (Activity) getContext());
    }

    public void showPageMainFrameError() {
        View container = this.mErrorLayout;
        if (container != null) {
            container.setVisibility(View.VISIBLE);
        } else {
            createErrorLayout();
            container = this.mErrorLayout;
        }
        View clickView = container.findViewById(mClickId);
        if (mClickId != -1 && clickView != null) {
            clickView.setClickable(true);
        } else {
            container.setClickable(true);
        }
    }

    private void createErrorLayout() {
        final FrameLayout mFrameLayout = new FrameLayout(getContext());
        mFrameLayout.setBackgroundColor(Color.WHITE);
        mFrameLayout.setId(R.id.mainframe_error_container_id);
        if (this.mErrorView == null) {
            LayoutInflater mLayoutInflater = LayoutInflater.from(getContext());
            Timber.i("mErrorLayoutRes:" + mErrorLayoutRes);
            mLayoutInflater.inflate(mErrorLayoutRes, mFrameLayout, true);
        } else {
            mFrameLayout.addView(mErrorView);
        }
        ViewStub mViewStub = this.findViewById(R.id.mainframe_error_viewsub_id);
        final int index = this.indexOfChild(mViewStub);
        this.removeViewInLayout(mViewStub);
        final ViewGroup.LayoutParams layoutParams = getLayoutParams();
        if (layoutParams != null) {
            this.addView(this.mErrorLayout = mFrameLayout, index, layoutParams);
        } else {
            this.addView(this.mErrorLayout = mFrameLayout, index);
        }
        mFrameLayout.setVisibility(View.VISIBLE);
        if (mClickId != -1) {
            final View clickView = mFrameLayout.findViewById(mClickId);
            if (clickView != null) {
                clickView.setOnClickListener(v -> {
                    if (getWebView() != null) {
                        clickView.setClickable(false);
                        getWebView().reload();
                    }
                });
                return;
            } else {
                Timber.e("ClickView is null , cannot bind accurate view to refresh or reload .");
            }
        }
        mFrameLayout.setOnClickListener(v -> {
            if (getWebView() != null) {
                mFrameLayout.setClickable(false);
                getWebView().reload();
            }
        });
    }

    public void hideErrorLayout() {
        View mView = null;
        if ((mView = this.findViewById(R.id.mainframe_error_container_id)) != null) {
            mView.setVisibility(View.GONE);
        }
    }

    public void setErrorView(@NonNull View errorView) {
        this.mErrorView = errorView;
    }

    public void setErrorLayoutRes(@LayoutRes int resLayout, @IdRes int id) {
        this.mClickId = id;
        if (this.mClickId <= 0) {
            this.mClickId = -1;
        }
        this.mErrorLayoutRes = resLayout;
        if (this.mErrorLayoutRes <= 0) {
            this.mErrorLayoutRes = R.layout.agentweb_error_page;
        }
    }

    @Override
    public AbsAgentWebUIController provide() {
        return this.mAgentWebUIController;
    }

    public void bindWebView(android.webkit.WebView view) {
        if (this.mWebView == null) {
            this.mWebView = view;
        }
    }

    public android.webkit.WebView getWebView() {
        return this.mWebView;
    }

    public void bindWebView(com.tencent.smtt.sdk.WebView view) {
        if (this.mX5WebView == null) {
            this.mX5WebView = view;
        }
    }

    public com.tencent.smtt.sdk.WebView getX5WebView() {
        return this.mX5WebView;
    }

}

File: scaffold/src/main/java/top/xuqingquan/web/publics/WebParentLayout.kt
package top.xuqingquan.web.publics

import android.app.Activity
import android.content.Context
import android.graphics.Color
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.ViewStub
import android.widget.FrameLayout
import androidx.annotation.IdRes
import androidx.annotation.LayoutRes
import top.xuqingquan.R
import top.xuqingquan.utils.Timber
import top.xuqingquan.web.nokernel.Provider

class WebParentLayout @JvmOverloads internal constructor(
    context: Context,
    attrs: AttributeSet?,
    defStyleAttr: Int = -1
) : FrameLayout(context, attrs, defStyleAttr), Provider<AbsAgentWebUIController> {
    private var mAgentWebUIController: AbsAgentWebUIController? = null
    @LayoutRes
    private var mErrorLayoutRes: Int = 0
    @IdRes
    private var mClickId = -1
    private var mErrorView: View? = null
    var webView: android.webkit.WebView? = null
        private set
    var x5WebView: com.tencent.smtt.sdk.WebView? = null
        private set
    private var mErrorLayout: FrameLayout? = null

    constructor(context: Context) : this(context, null) {
        Timber.i("WebParentLayout")
    }

    init {
        if (context !is Activity) {
            throw IllegalArgumentException("WebParentLayout context must be activity or activity sub class .")
        }
        this.mErrorLayoutRes = R.layout.agentweb_error_page
    }

    fun bindController(agentWebUIController: AbsAgentWebUIController) {
        this.mAgentWebUIController = agentWebUIController
        this.mAgentWebUIController!!.bindWebParent(this, context as Activity)
    }

    fun showPageMainFrameError() {
        var container: View? = this.mErrorLayout
        if (container != null) {
            container.visibility = View.VISIBLE
        } else {
            createErrorLayout()
            container = this.mErrorLayout
        }
        val clickView = container!!.findViewById<View>(mClickId)
        if (mClickId != -1 && clickView != null) {
            clickView.isClickable = true
        } else {
            container.isClickable = true
        }
    }

    private fun createErrorLayout() {
        val mFrameLayout = FrameLayout(context)
        mFrameLayout.setBackgroundColor(Color.WHITE)
        mFrameLayout.id = R.id.mainframe_error_container_id
        if (this.mErrorView == null) {
            val mLayoutInflater = LayoutInflater.from(context)
            Timber.i("mErrorLayoutRes:$mErrorLayoutRes")
            mLayoutInflater.inflate(mErrorLayoutRes, mFrameLayout, true)
        } else {
            mFrameLayout.addView(mErrorView)
        }
        val mViewStub = this.findViewById<ViewStub>(R.id.mainframe_error_viewsub_id)
        val index = this.indexOfChild(mViewStub)
        this.removeViewInLayout(mViewStub)
        val layoutParams = layoutParams
        this.mErrorLayout = mFrameLayout
        if (layoutParams != null) {
            this.addView(mFrameLayout, index, layoutParams)
        } else {
            this.addView(mFrameLayout, index)
        }
        mFrameLayout.visibility = View.VISIBLE
        if (mClickId != -1) {
            val clickView = mFrameLayout.findViewById<View>(mClickId)
            if (clickView != null) {
                clickView.setOnClickListener { v ->
                    if (webView != null) {
                        clickView.isClickable = false
                        webView!!.reload()
                    }
                }
                return
            } else {
                Timber.e("ClickView is null , cannot bind accurate view to refresh or reload .")
            }
        }
        mFrameLayout.setOnClickListener {
            if (webView != null) {
                mFrameLayout.isClickable = false
                webView!!.reload()
            }
        }
    }

    fun hideErrorLayout() {
        val mView = this.findViewById<View>(R.id.mainframe_error_container_id)
        mView?.visibility = View.GONE
    }

    fun setErrorView(errorView: View?) {
        this.mErrorView = errorView
    }

    fun setErrorLayoutRes(@LayoutRes resLayout: Int, @IdRes id: Int) {
        this.mClickId = id
        if (this.mClickId <= 0) {
            this.mClickId = -1
        }
        this.mErrorLayoutRes = resLayout
        if (this.mErrorLayoutRes <= 0) {
            this.mErrorLayoutRes = R.layout.agentweb_error_page
        }
    }

    override fun provide(): AbsAgentWebUIController? {
        return this.mAgentWebUIController
    }

    fun bindWebView(view: android.webkit.WebView) {
        if (this.webView == null) {
            this.webView = view
        }
    }

    fun bindWebView(view: com.tencent.smtt.sdk.WebView) {
        if (this.x5WebView == null) {
            this.x5WebView = view
        }
    }

}
